crates/wasmtime/src/component/types.rs

```diff
@@ -0,0 +1,1023 @@
+                index: *index,
+                types: types.clone(),
+            })),
+            InterfaceType::Flags(index) => Type::Flags(Flags(Handle {
+                index: *index,
+                types: types.clone(),
+            })),
+        }
+    }
+
+    /// Return the number of stack slots needed to store values of this type in lowered form.
+    pub(crate) fn flatten_count(&self) -> usize {
+        match self {
+            Type::Unit => 0,
+
+            Type::Bool
+            | Type::S8
+            | Type::U8
+            | Type::S16
+            | Type::U16
+            | Type::S32
+            | Type::U32
+            | Type::S64
+            | Type::U64
+            | Type::Float32
+            | Type::Float64
+            | Type::Char
+            | Type::Enum(_) => 1,
+
+            Type::String | Type::List(_) => 2,
+
+            Type::Record(handle) => handle.fields().map(|field| field.ty.flatten_count()).sum(),
+
+            Type::Tuple(handle) => handle.types().map(|ty| ty.flatten_count()).sum(),
+
+            Type::Variant(handle) => {
+                1 + handle
+                    .cases()
+                    .map(|case| case.ty.flatten_count())
+                    .max()
+                    .unwrap_or(0)
+            }
+
+            Type::Union(handle) => {
+                1 + handle
+                    .types()
+                    .map(|ty| ty.flatten_count())
+                    .max()
+                    .unwrap_or(0)
+            }
+
+            Type::Option(handle) => 1 + handle.ty().flatten_count(),
+
+            Type::Expected(handle) => {
+                1 + handle
+                    .ok()
+                    .flatten_count()
+                    .max(handle.err().flatten_count())
+            }
+
+            Type::Flags(handle) => values::u32_count_for_flag_count(handle.names().len()),
+        }
+    }
+
+    fn desc(&self) -> &'static str {
+        match self {
+            Type::Unit => "unit",
+            Type::Bool => "bool",
+            Type::S8 => "s8",
+            Type::U8 => "u8",
+            Type::S16 => "s16",
+            Type::U16 => "u16",
+            Type::S32 => "s32",
+            Type::U32 => "u32",
+            Type::S64 => "s64",
+            Type::U64 => "u64",
+            Type::Float32 => "float32",
+            Type::Float64 => "float64",
+            Type::Char => "char",
+            Type::String => "string",
+            Type::List(_) => "list",
+            Type::Record(_) => "record",
+            Type::Tuple(_) => "tuple",
+            Type::Variant(_) => "variant",
+            Type::Enum(_) => "enum",
+            Type::Union(_) => "union",
+            Type::Option(_) => "option",
+            Type::Expected(_) => "expected",
+            Type::Flags(_) => "flags",
+        }
+    }
+
+    /// Deserialize a value of this type from core Wasm stack values.
+    pub(crate) fn lift<'a>(
```

Comment:
Would it be possible to move `lift`/`load` to the `Val` module and keep all the internals of all the `Val` private (e.g. not `pub(crate)`)?